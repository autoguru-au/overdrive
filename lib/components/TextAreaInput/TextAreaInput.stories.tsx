import { action } from '@storybook/addon-actions';
import { ArgTypes, Meta, StoryFn } from '@storybook/react';
import * as React from 'react';
import { ComponentProps } from 'react';

import { TextAreaInput } from '.';

export default {
	title: 'Components/Inputs/Textarea',
	component: TextAreaInput,
	parameters: { chromatic: {} },
} satisfies Meta<typeof TextAreaInput>;

const defaultValue =
	'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus.\nPhasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc, quis gravida magna mi a libero. Fusce vulputate eleifend sapien. Vestibulum purus quam, scelerisque ut, mollis sed, nonummy id, metus. Nullam accumsan lorem in dui. Cras ultricies mi eu turpis hendrerit fringilla. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; In ac dui quis mi consectetuer lacinia. Nam pretium turpis et arcu. Duis arcu tortor, suscipit eget, imperdiet nec, imperdiet iaculis, ipsum. Sed aliquam ultrices mauris. Integer ante arcu, accumsan a, consectetuer eget, posuere ut, mauris. Praesent adipiscing. Phasellus ullamcorper ipsum rutrum nunc. Nunc nonummy metus. Vestibulum volutpat pretium libero. Cras id dui. Aenean ut';
const defaultPlaceholder = 'Tell us about your car.';

const argTypes: ArgTypes = {
	maxLength: {
		defaultValue: null,
		description: 'Set the max length of the input',
		control: {
			type: 'number',
		},
	},
};

const Template: StoryFn<typeof TextAreaInput> = (args) => (
	<TextAreaInput {...args} />
);

const sharedProps: ComponentProps<typeof TextAreaInput> = {
	disabled: false,
	name: 'text',
	placeholder: defaultPlaceholder,
	isValid: false,
	isTouched: false,
	isLoading: false,
	isFocused: false,
	reserveHintSpace: false,
	hintText: '',
	notch: true,
	prefixIcon: void 0,
	onChange: action('onChange'),
	onFocus: action('onFocus'),
	onBlur: action('onBlur'),
};

const standardProps: ComponentProps<typeof TextAreaInput> = sharedProps;
const withAValueProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	value: defaultValue,
	placeholder: defaultPlaceholder,
};
const withHintTextProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	hintText: 'Hint Text',
	placeholder: defaultPlaceholder,
};
const disabledProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	value: defaultValue,
	placeholder: defaultPlaceholder,
	disabled: true,
};
const validProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	value: defaultValue,
	placeholder: defaultPlaceholder,
	isTouched: true,
	isValid: true,
};
const invalidProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	value: 'Bob the builder',
	placeholder: defaultPlaceholder,
	isTouched: true,
	isValid: false,
	hintText: 'Cannot be Bob the builder',
};
const noNotchProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	placeholder: defaultPlaceholder,
	notch: false,
};
const noNotchWithValueProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	value: defaultValue,
	placeholder: defaultPlaceholder,
	notch: false,
};
const loadingProps: ComponentProps<typeof TextAreaInput> = {
	...sharedProps,
	isLoading: true,
};
export const standard = Template.bind(standardProps);
standard.args = standardProps;
standard.argTypes = argTypes;

export const withAValue = Template.bind(withAValueProps);
withAValue.args = withAValueProps;
withAValue.argTypes = argTypes;

export const withHintText = Template.bind(withHintTextProps);
withHintText.args = withHintTextProps;
withHintText.argTypes = argTypes;

export const notchDisabled = Template.bind(noNotchProps);
notchDisabled.args = noNotchProps;
notchDisabled.argTypes = argTypes;

export const notchDisabledWithValue = Template.bind(noNotchWithValueProps);
notchDisabledWithValue.args = noNotchWithValueProps;
notchDisabledWithValue.argTypes = argTypes;

export const disabled = Template.bind(disabledProps);
disabled.args = disabledProps;
disabled.argTypes = argTypes;

export const valid = Template.bind(validProps);
valid.args = validProps;
valid.argTypes = argTypes;

export const invalid = Template.bind(invalidProps);
invalid.args = invalidProps;
invalid.argTypes = argTypes;

export const loading = Template.bind(loadingProps);
loading.args = loadingProps;
loading.argTypes = argTypes;

const smallProps: typeof standardProps = {
	...standardProps,
	size: 'small',
};
export const small = Template.bind(smallProps);
small.args = smallProps;
small.argTypes = argTypes;

const withValueSmallProps: typeof withAValueProps = {
	...withAValueProps,
	size: 'small',
};

export const withValueSmall = Template.bind(withValueSmallProps);
withValueSmall.args = withValueSmallProps;
withValueSmall.argTypes = argTypes;

const loadingSmallProps: typeof withAValueProps = {
	...withAValueProps,
	isLoading: true,
	size: 'small',
};
export const loadingSmall = Template.bind(loadingSmallProps);
loadingSmall.args = loadingSmallProps;
loadingSmall.argTypes = argTypes;
