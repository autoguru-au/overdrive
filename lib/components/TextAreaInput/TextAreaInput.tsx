import * as React from 'react';

import { Box } from '../Box';
import { withEnhancedInput } from '../private/InputBase';

export const TextAreaInput = withEnhancedInput<{}, HTMLTextAreaElement>(
	({ field, eventHandlers, validation, suffixed, prefixed, ...rest }) => (
		<Box
			is="textarea"
			{...eventHandlers}
			{...field}
			{...rest}
			autoComplete="off"
		/>
	),
	{
		primitiveType: 'textarea',
		withSuffixIcon: false,
		withPrefixIcon: false,
	},
);
