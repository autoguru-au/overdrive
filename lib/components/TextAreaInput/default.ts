export { TextAreaInput as default } from './TextAreaInput';
