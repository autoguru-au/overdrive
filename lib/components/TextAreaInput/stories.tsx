import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import { TextAreaInput } from '.';

const sharedKnobs = placeholder => ({
	placeholder: text('Placeholder', placeholder),
	disabled: boolean('disabled', false),
	onChange: action('onChange'),
	onFocus: action('onFocus'),
	onBlur: action('onBlur'),
});

storiesOf('Components|Inputs/Textarea', module)
	.addParameters({
		chromatic: { delay: 300 },
	})
	.add('default', () => (
		<TextAreaInput name="abc" {...sharedKnobs('Tell us about your car.')} />
	))
	.add('with a value', () => (
		<TextAreaInput
			name="abc"
			placeholder="Tell us about your car."
			value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vitae pulvinar odio. Duis laoreet lacus vel consequat congue. Ut euismod enim non eros lacinia mollis. Vestibulum libero quam, aliquet non justo laoreet, egestas molestie ante. Quisque urna leo, consectetur id dui aliquet, placerat iaculis augue. Pellentesque sed vestibulum augue, quis porta lectus."
		/>
	))
	.add('with hint text', () => (
		<TextAreaInput
			name="abc"
			placeholder="How many?"
			hintText={text('Hint Text', 'Must be greater than 10.')}
		/>
	))
	.add('disabled', () => (
		<TextAreaInput
			disabled
			name="abc"
			placeholder="Tell us about your car."
			value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vitae pulvinar odio. Duis laoreet lacus vel consequat congue. Ut euismod enim non eros lacinia mollis. Vestibulum libero quam, aliquet non justo laoreet, egestas molestie ante. Quisque urna leo, consectetur id dui aliquet, placerat iaculis augue. Pellentesque sed vestibulum augue, quis porta lectus."
		/>
	));
// TODO: A test here to limit the value to 100 chars
