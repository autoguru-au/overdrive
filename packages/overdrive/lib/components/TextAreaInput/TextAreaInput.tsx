import * as React from 'react';

import { Box } from '../Box';
import { withEnhancedInput } from '../private/InputBase';

const MAX_TEXT_AREA_INPUT_LENGTH = 4000;

export const TextAreaInput = withEnhancedInput<
	Partial<Pick<HTMLTextAreaElement, 'type' | 'maxLength'>>
>(
	({
		field,
		eventHandlers,
		validation,
		isLoading,
		suffixed,
		prefixed,
		size,
		...rest
	}) => (
		<Box
			is="textarea"
			{...eventHandlers}
			{...field}
			{...rest}
			autoComplete="off"
			maxLength={rest.maxLength || MAX_TEXT_AREA_INPUT_LENGTH}
		/>
	),
	{
		primitiveType: 'textarea',
		withSuffixIcon: false,
		withPrefixIcon: false,
	},
);

export default TextAreaInput;
