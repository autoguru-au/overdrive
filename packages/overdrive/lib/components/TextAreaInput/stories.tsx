import { text } from '@storybook/addon-knobs';
import * as React from 'react';

import { TextAreaInput } from '.';

export default {
	title: 'Components/Inputs/Textarea',
	component: TextAreaInput,
	parameters: {
		chromatic: { delay: 300 },
	},
};

export const standard = () => (
	<TextAreaInput name="abc" placeholder="Tell us about your car." />
);
export const withAValue = () => (
	<TextAreaInput
		name="abc"
		placeholder="Tell us about your car."
		value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vitae pulvinar odio. Duis laoreet lacus vel consequat congue. Ut euismod enim non eros lacinia mollis. Vestibulum libero quam, aliquet non justo laoreet, egestas molestie ante. Quisque urna leo, consectetur id dui aliquet, placerat iaculis augue. Pellentesque sed vestibulum augue, quis porta lectus."
	/>
);
export const withHintText = () => (
	<TextAreaInput
		name="abc"
		placeholder="How many?"
		hintText={text('Hint Text', 'Must be greater than 10.')}
	/>
);
export const disabled = () => (
	<TextAreaInput
		disabled
		name="abc"
		placeholder="Tell us about your car."
		value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vitae pulvinar odio. Duis laoreet lacus vel consequat congue. Ut euismod enim non eros lacinia mollis. Vestibulum libero quam, aliquet non justo laoreet, egestas molestie ante. Quisque urna leo, consectetur id dui aliquet, placerat iaculis augue. Pellentesque sed vestibulum augue, quis porta lectus."
	/>
);
export const loading = () => (
	<TextAreaInput
		isLoading
		name="abc"
		placeholder="Tell us about your car."
		value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vitae pulvinar odio. Duis laoreet lacus vel consequat congue. Ut euismod enim non eros lacinia mollis. Vestibulum libero quam, aliquet non justo laoreet, egestas molestie ante. Quisque urna leo, consectetur id dui aliquet, placerat iaculis augue. Pellentesque sed vestibulum augue, quis porta lectus."
	/>
);
// TODO: A test here to limit the value to 100 chars
