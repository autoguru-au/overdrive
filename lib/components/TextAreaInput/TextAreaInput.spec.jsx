import React from 'react';
import { mount, render } from 'enzyme';
import { TextAreaInput } from './TextAreaInput';

describe('<TextAreaInput />', () => {
	it('should not throw', () => {
		expect(() =>
			mount(
				<TextAreaInput placeholder="placeholder something" />,
			).unmount(),
		).not.toThrow();
	});

	it('should have a textarea tag', () => {
		const textAreaInput = mount(
			<TextAreaInput placeholder="placeholder something" />,
		);
		expect(textAreaInput.find('textarea').exists()).toEqual(true);
		textAreaInput.unmount();
	});

	it('should pass on className to dom element', () => {
		const textAreaInput = mount(
			<TextAreaInput
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);
		expect(textAreaInput.hasClass('textarea-input-class')).toBeTruthy();
		textAreaInput.unmount();
	});

	it('should match snapshot for default select input', () => {
		const textAreaInput = render(
			<TextAreaInput
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);
		expect(textAreaInput).toMatchSnapshot();
	});

	it('should match snapshot for active select input', () => {
		const textAreaInput = mount(
			<TextAreaInput
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);

		textAreaInput.find('textarea').simulate('focus');

		textAreaInput
			.find('textarea')
			.simulate('change', { target: { value: 'Test value' } });

		// Snapshot confirms this
		expect(textAreaInput).toMatchSnapshot();

		textAreaInput.unmount();
	});

	it('should match snapshot for touched select input', () => {
		const textAreaInput = render(
			<TextAreaInput
				isTouched
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);
		expect(textAreaInput).toMatchSnapshot();
	});

	it('should match snapshot for touched valid select input', () => {
		const textAreaInput = render(
			<TextAreaInput
				isTouched
				isValid
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);
		expect(textAreaInput).toMatchSnapshot();
	});

	it('should match snapshot for touched invalid select input', () => {
		const textAreaInput = render(
			<TextAreaInput
				isTouched
				isValid={false}
				className="textarea-input-class"
				placeholder="placeholder something"
			/>,
		);
		expect(textAreaInput).toMatchSnapshot();
	});

	it('should display placeholder text', () => {
		const textAreaInput = render(
			<TextAreaInput placeholder="placeholder something" />,
		);
		expect(textAreaInput.find('label').text()).toEqual(
			'placeholder something',
		);
	});

	it('should fire onFocus event', () => {
		const spyedCallback = jest.fn();
		const textAreaInput = mount(
			<TextAreaInput
				placeholder="placeholder something"
				onFocus={spyedCallback}
			/>,
		);
		textAreaInput.find('textarea').simulate('focus');

		expect(spyedCallback).toHaveBeenCalled();
		textAreaInput.unmount();
	});

	it('should fire onBlur event', () => {
		const spyedCallback = jest.fn();
		const textAreaInput = mount(
			<TextAreaInput
				placeholder="placeholder something"
				onBlur={spyedCallback}
			/>,
		);
		textAreaInput.find('textarea').simulate('blur');

		expect(spyedCallback).toHaveBeenCalled();
		textAreaInput.unmount();
	});

	it('should fire onChange event', () => {
		const spyedCallback = jest.fn();
		const textAreaInput = mount(
			<TextAreaInput
				placeholder="placeholder something"
				onChange={spyedCallback}
			/>,
		);
		textAreaInput.find('textarea').simulate('change');

		expect(spyedCallback).toHaveBeenCalled();
		textAreaInput.unmount();
	});
});
