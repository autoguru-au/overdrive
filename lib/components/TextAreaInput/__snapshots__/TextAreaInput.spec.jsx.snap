// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAreaInput /> should match snapshot for active select input 1`] = `
<EnhancedInput(TextAreaInput)
  className="textarea-input-class"
  placeholder="placeholder something"
>
  <div
    className="root textarea-input-class"
  >
    <NotchedBase
      isActive={true}
      isEmpty={false}
      placeholder="placeholder something"
    >
      <div
        className="notchedBase notchedBaseShift notchedBaseActive"
      >
        <TextAreaInput
          eventHandlers={
            Object {
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
            }
          }
          field={
            Object {
              "className": "input",
              "disabled": false,
              "id": undefined,
              "name": undefined,
              "value": "Test value",
            }
          }
          validation={
            Object {
              "isTouched": undefined,
              "isValid": undefined,
            }
          }
        >
          <textarea
            autoComplete="off"
            className="input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value="Test value"
          />
        </TextAreaInput>
        <div
          className="notchedBaseBorder"
        >
          <div
            className="notchedBaseBorderLeading"
          />
          <div
            className="notchedBaseBorderNotch"
            style={
              Object {
                "width": 8,
              }
            }
          >
            <label
              className="notchedBasePlaceholder"
            >
              placeholder something
            </label>
          </div>
          <div
            className="notchedBaseBorderTrailing"
          />
        </div>
      </div>
    </NotchedBase>
  </div>
</EnhancedInput(TextAreaInput)>
`;

exports[`<TextAreaInput /> should match snapshot for default select input 1`] = `
<div
  class="root textarea-input-class"
>
  <div
    class="notchedBase"
  >
    <textarea
      autocomplete="off"
      class="input"
    />
    <div
      class="notchedBaseBorder"
    >
      <div
        class="notchedBaseBorderLeading"
      />
      <div
        class="notchedBaseBorderNotch"
        style="width:0"
      >
        <label
          class="notchedBasePlaceholder"
        >
          placeholder something
        </label>
      </div>
      <div
        class="notchedBaseBorderTrailing"
      />
    </div>
  </div>
</div>
`;

exports[`<TextAreaInput /> should match snapshot for touched invalid select input 1`] = `
<div
  class="root textarea-input-class invalid withStatus"
>
  <div
    class="notchedBase"
  >
    <textarea
      autocomplete="off"
      class="input"
    />
    <div
      class="notchedBaseBorder"
    >
      <div
        class="notchedBaseBorderLeading"
      />
      <div
        class="notchedBaseBorderNotch"
        style="width:0"
      >
        <label
          class="notchedBasePlaceholder"
        >
          placeholder something
        </label>
      </div>
      <div
        class="notchedBaseBorderTrailing"
      />
    </div>
  </div>
</div>
`;

exports[`<TextAreaInput /> should match snapshot for touched select input 1`] = `
<div
  class="root textarea-input-class"
>
  <div
    class="notchedBase"
  >
    <textarea
      autocomplete="off"
      class="input"
    />
    <div
      class="notchedBaseBorder"
    >
      <div
        class="notchedBaseBorderLeading"
      />
      <div
        class="notchedBaseBorderNotch"
        style="width:0"
      >
        <label
          class="notchedBasePlaceholder"
        >
          placeholder something
        </label>
      </div>
      <div
        class="notchedBaseBorderTrailing"
      />
    </div>
  </div>
</div>
`;

exports[`<TextAreaInput /> should match snapshot for touched valid select input 1`] = `
<div
  class="root textarea-input-class valid withStatus"
>
  <div
    class="notchedBase"
  >
    <textarea
      autocomplete="off"
      class="input"
    />
    <div
      class="notchedBaseBorder"
    >
      <div
        class="notchedBaseBorderLeading"
      />
      <div
        class="notchedBaseBorderNotch"
        style="width:0"
      >
        <label
          class="notchedBasePlaceholder"
        >
          placeholder something
        </label>
      </div>
      <div
        class="notchedBaseBorderTrailing"
      />
    </div>
  </div>
</div>
`;
